### ISSUE TITLE
Performance regression with vector strategies

### ISSUE THREAD

**Issue by sabauma** [samth/pycket#197]
Vector strategies are resulting in a non-trivial amount of extra work being performed when using an in [Racket implementation of HAMTs](https://github.com/97jaz/hamt). The following seems to be a common idiom.

```racket
(define src (vector holding the entries for the current node))
(define dst (make-vector (add1 (vector-length src)) 0))
(vector-copy! dst 0 src)
```

I suspect part of our problem is poor hash functions resulting in very large vectors representing hash-collision nodes. Even if that is the case, this still performs unwanted copying and allocation.
Currently, Pycket will allocate an unboxed array of 0s in `make-vector`.  Invoking `vector-copy!` immediately reallocates the `dst` vector and boxes each (identical) fixnum. Pycket allocates significantly more than Racket on this code.

**Comment by samth** [comment]
Can't we just make `vector-copy!` smarter so that it uses the fixnum strategy?

**Comment by sabauma** [comment]
I think there are a couple possible strategies for addressing this problem, of varying degrees of generality.

1. Rework `make-vector` to use the generic object strategy. This avoids the need to respecialize in `vector-copy!` and avoids the allocation of a large number of identical fixnum objects. While easy to implement, this disables specialization of anything created by `make-vector`.
2. Rework `vector-copy!` to be a bit smarter in certain cases. This is tricky as Pycket has to interact with the evaluator in the general case, but the common case for `vector-copy!` is two unchaperoned vectors.
3. Create a `make-vector` strategy which has no backing array storage, just its size and a pointer to the default element. This is probably the preferable solution.
4. Something I haven't come up with yet.

**Comment by sabauma** [comment]
@samth: The problem is that it can't use the fixnum strategy, we have to switch to a general object strategy, as the `src` vector is a vector of structs.

**Comment by samth** [comment]
I think we should do both 2 and 3. 

One other thing to think about when doing 3: this idiom ought to be made faster than it is in Pycket:

```
(define v (make-vector 1000 #f))
(for ([i 1000]) (vector-set! v 0.1))
```

Right now, this produces a vector of floats using the object strategy. It would be nice if we could avoid that.

**Comment by samth** [comment]
Also, `vector-copy!` could be smarter in the general case by not filling in slots that will be overridden.

**Comment by sabauma** [comment]
This was mostly fixed by ff37c589951778e642690eb6ef47eced27dc12d9. This not only saves additional array allocation, but prevents us from generating a large number of identical fixnum objects during specialization.

With other recent fixes, this brings performance under 2x slowdown vs Racket using the HAMT library. I think the remaining overhead is due to a suboptimal `vector-copy!`, which does not seem to JIT nicely in some cases.
